File: src/com/example/javas/medium/M021_MinimumWindowSubsequence.java
package com.example.javas.medium;

/**
 * M021: Minimum Window Subsequence
 * Difficulty: Medium
 * Time Target: 25 minutes
 * 
 * Problem:
 * Given strings s and t, find the minimum (shortest) substring of s that contains 
 * all characters of t in the same order (as a subsequence).
 * Return the minimum window. If there is no such window, return an empty string "".
 * 
 * Requirements:
 * - Find shortest substring of s where t is a subsequence
 * - Characters in t must appear in order in the substring
 * - If multiple windows exist, return any one
 * - If no valid window exists, return ""
 * - 1 <= s.length <= 2 * 10^4
 * - 1 <= t.length <= 100
 * - s and t consist of lowercase English letters
 * 
 * Examples:
 * Input: s = "abcdebdde", t = "bde"
 * Output: "bcde"
 * Explanation: "bcde" is the shortest substring where 'b', 'd', 'e' appear in order.
 * Note: "bdde" or "bde" would also be valid answers.
 * 
 * Input: s = "jmeqksfrsdcmsiwvaovztaqenprpvnbstl", t = "u"
 * Output: ""
 * Explanation: 'u' does not exist in s.
 * 
 * Input: s = "abcde", t = "ace"
 * Output: "abcde"
 * Explanation: The entire string contains 'a', 'c', 'e' in order.
 * 
 * Input: s = "fgrqsqsnodwmxzkzxwqegkndaa", t = "kzed"
 * Output: "kzxwqegknd"
 * 
 * Approach:
 * Use two-pointer technique with forward matching and backward shrinking:
 * 1. Forward pass: Use two pointers (i for s, j for t) to find window end
 *    - Match each character in t sequentially by advancing i in s
 *    - When all characters matched, we have a valid window
 * 2. Backward pass: Shrink window from left to find minimum
 *    - Start from matched position and go backwards
 *    - Match characters of t in reverse order to find optimal start
 * 3. Track minimum window and repeat from next position
 * 
 * Example: s = "abcdebdde", t = "bde"
 * - Forward: match b(1), d(3), e(4) -> window "bcde" 
 * - Backward: shrink to find if shorter window exists
 * - Continue searching for other possible windows
 * 
 * Data Structures:
 * - Two pointers for tracking positions in s and t
 * - Variables to track minimum window start and length
 * 
 * Time Complexity: O(s * t) in worst case
 * Space Complexity: O(1) - only using pointers and variables
 */
public class M021_MinimumWindowSubsequence {
    
    public static String minWindow(String s, String t) {
        // TODO: Implement minimum window subsequence finder
        // 1. Initialize: minLen = Integer.MAX_VALUE, minStart = 0, sIndex = 0
        // 2. While sIndex < s.length():
        //    a) Forward pass - find window end:
        //       - Use tIndex to track position in t
        //       - Advance sIndex in s until all characters in t are matched
        //       - If can't match all of t, break (no more valid windows)
        //    b) Backward pass - shrink window to find minimum:
        //       - Save end position (sIndex - 1)
        //       - Go backwards from end, matching t in reverse
        //       - Find the leftmost position that still contains subsequence
        //    c) Update minimum window if current window is smaller
        //    d) Move sIndex to position after window start for next iteration
        // 3. Return substring if minLen was updated, otherwise return ""
        
        return "";
    }
    
    
    // Test cases
    public static void main(String[] args) {
        // Test case 1: Basic case
        String s1 = "abcdebdde";
        String t1 = "bde";
        System.out.println("Test 1: \"" + minWindow(s1, t1) + "\""); // Expected: "bcde" or "bdde" or "bde"
        
        // Test case 2: Character not found
        String s2 = "jmeqksfrsdcmsiwvaovztaqenprpvnbstl";
        String t2 = "u";
        System.out.println("Test 2: \"" + minWindow(s2, t2) + "\""); // Expected: ""
        
        // Test case 3: Entire string needed
        String s3 = "abcde";
        String t3 = "ace";
        System.out.println("Test 3: \"" + minWindow(s3, t3) + "\""); // Expected: "abcde"
        
        // Test case 4: Complex case
        String s4 = "fgrqsqsnodwmxzkzxwqegkndaa";
        String t4 = "kzed";
        System.out.println("Test 4: \"" + minWindow(s4, t4) + "\""); // Expected: "kzxwqegknd"
        
        // Test case 5: Single character match
        String s5 = "abc";
        String t5 = "c";
        System.out.println("Test 5: \"" + minWindow(s5, t5) + "\""); // Expected: "c"
    }
}

File: pythonX/medium/m021_minimum_window_subsequence.py
"""
M021: Minimum Window Subsequence

Summary:
Find shortest substring of s where t is a subsequence.

Requirements:
Find minimum window containing all characters of t in order.

Approach:
Two-pointer with forward matching and backward shrinking.

Data Structures:
- String pointers

Big O Analysis:
Time Complexity: O(s * t)
Space Complexity: O(1)
"""


def min_window(s: str, t: str) -> str:
    """
    TODO: Find minimum window subsequence
    1. Initialize min_len, min_start, s_index
    2. While s_index < len(s):
       a) Forward pass - find window end:
          - Match all chars in t sequentially
          - If can't match, break
       b) Backward pass - shrink to find minimum:
          - Go backwards matching t in reverse
       c) Update minimum if smaller
       d) Move s_index forward
    3. Return substring if found, else ""
    """
    pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Basic case")
    print(f"'abcdebdde', 'bde': '{min_window('abcdebdde', 'bde')}'")  # Expected: bcde or bdde or bde
    
    print("\nTest 2: Character not found")
    print(f"'jmeqksfrsdcmsiwvaovztaqenprpvnbstl', 'u': '{min_window('jmeqksfrsdcmsiwvaovztaqenprpvnbstl', 'u')}'")  # Expected: ""
    
    print("\nTest 3: Entire string needed")
    print(f"'abcde', 'ace': '{min_window('abcde', 'ace')}'")  # Expected: abcde
    
    print("\nTest 4: Complex case")
    print(f"'fgrqsqsnodwmxzkzxwqegkndaa', 'kzed': '{min_window('fgrqsqsnodwmxzkzxwqegkndaa', 'kzed')}'")  # Expected: kzxwqegknd
    
    print("\nTest 5: Single character")
    print(f"'abc', 'c': '{min_window('abc', 'c')}'")  # Expected: c
